hedera-node/docs/design/services/smart-contract-service/IHFC-associate-dissociate.md

```diff
@@ -0,0 +1,81 @@
+# IHRC Facade Associate and Dissociate Functionality
+
+## Purpose
+
+[HIP-218](https://hips.hedera.com/hip/hip-218) proposed a way to enable treating HTS tokens in an equivalent manner
+to ERC tokens in the EVM world by creating a proxy redirect facade contract.  We need to extend 
+this functionality to include `associate` and `dissociate` functions which are part of functionality available to HTS tokens
+but not ERC tokens.
+
+## Goals
+
+- Extend the proxy redirect token facade contract to enable calling `associate` and `dissociate` functions on tokens.
+
+## Non Goals
+
+- Refactor the current implementation to better support future additions to the token facade contract functionality.
+
+## Architecture
+
+The method in which the redirection to proxy contract is implemented is that during the execution of an EVM request, we examine the contract address to see
+if it is an HTS token address and if so we redirect the request to the proxy contract.  This mechanism will be extended to include
+calls to `associate` and `dissociate` functions in the HTSPrecompileContact class method which handles ABI_ID_REDIRECT_FOR_TOKEN function selector.
+
+The following table describes the function selector for the new `associate` and `dissociate` functions and the associated function signatures.
+
+| Function Selector | Function Signature |
+|-------------------|--------------------|
+| 0x0a754de6        | associate()        |
+| 0x5c9217e0        | dissociate()       |
+
+No arguments are necessary because 
+- The token address was already determined by looking up the contract address to determine if it is an HTS token address.
+- The address to associate/dissociate will be the caller (i.e. msg.sender)
+
+Once the above functionality has been implemented in services, the end user will be able to call the `associate` and `dissociate` functions as follows:
+
+```
+IHRC(tokenAddress).associate()
+IHRC(tokenAddress).dissociate()
+```
+
+The solidity interface for IHRC will be the following
+
+```
+interface IHRC {
+    function associate() external returns (responseCode);
+    function dissociate() external returns (responseCode);
+}
+```
+
+## Implementation
+
+Override the existing `AbstractAssociatePrecompile` and `AbstractDissociatePrecompile` classes to handle this new use case
+of being called via the proxy contract.  The new overwritten classes will differ from the existing classes in that they will
+be passed the token id and the sender address as constructor arguments and construct the associate and dissociate operator 
+classes from these arguments.  It will also need to send a value into the constructor of the parent class to signal that
+signature checking will not be required. 
```

Comment:
It may be dangerous to add such a flag for whether signature checks are required. I don't think we'll need to skip signature verifications. There is a rule in the signature verification logic stating that if the `sender` is the account that needs to authorize the transaction, we consider it valid and signed. This means that we don't need to skip the verifications, they will be successful.

Reply:
If there's a way to make this more explicit and maybe separate this may make refactoring/removal of grandfathering logic cleaner and safer.

It'd be nice to have an explicit check that confirms msg.sender is caller or that sender is the entity whose state will be modified.

Reply:
I believe that [this](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/sources/TxnAwareEvmSigsVerifier.java#L241) check ensures that the only way through the signature check in the new verification logic will succeed is if caller is the entity whose state is being modified.